src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -1060,12 +1203,61 @@ struct BlockZipperJoinImpl {
   }
 
   // The actual join routine that combines all the previous functions.
+  // not been consumed yet and match them against undef blocks.
+  template <bool reversed>
+  bool consumeRemainingBlocks(auto& side, const auto& undefBlocks) {
+    bool hasUndef = false;
+    while (side.it_ != side.end_) {
+      const auto& lBlock = *side.it_;
+      for (const auto& rBlock : undefBlocks) {
+        if constexpr (reversed) {
+          compatibleRowAction_.setInput(rBlock.fullBlock(), lBlock);
+        } else {
+          compatibleRowAction_.setInput(lBlock, rBlock.fullBlock());
+        }
+        for (size_t i : ad_utility::integerRange(lBlock.size())) {
+          for (size_t j : rBlock.getIndexRange()) {
+            hasUndef = true;
+            if constexpr (reversed) {
+              compatibleRowAction_.addRow(j, i);
+            } else {
+              compatibleRowAction_.addRow(i, j);
+            }
+          }
+        }
+      }
+      ++side.it_;
+    }
+    return hasUndef;
+  }
+
+  // If one of the sides is exhausted and has no values to match the pairs
+  // left, we need to pair the remaining values with the undef values we have
+  // left.
+  void addRemainingUndefPairs() {
+    if constexpr (potentiallyHasUndef) {
+      addCartesianProduct(leftSide_.currentBlocks_, rightSide_.undefBlocks_);
+      consumeRemainingBlocks<false>(leftSide_, rightSide_.undefBlocks_);
+
+      addCartesianProduct(leftSide_.undefBlocks_, rightSide_.currentBlocks_);
+      consumeRemainingBlocks<true>(rightSide_, leftSide_.undefBlocks_);
```

Comment:
Code duplication between left and right:)
And can be `if ! undef return` (less indentation).

Reply:
I think the code duplication is acceptable in this case. Adding a helper function with template parameters and stuff just because we're calling 2 functions per side instead of one makes things harder to read in my opinion.

Also the `if constexpr` could be flipped, but if I'm not mistaken this will cause the compiler to instantiate the functions and rely on an optimization pass to eliminate the dead code, whereas this way it won't even try to semantically compile it.